File: misc/2024/day24.jl
#!/usr/bin/env julia
# using Pkg
# Pkg.add("DrawSimpleGraphs")
# Pkg.add("JSON")
# Pkg.add("Plots")
# Pkg.add("SimpleGraphs")
using DrawSimpleGraphs, JSON, Plots, SimpleGraphs

# Load the JSON file
json_path = "day24.json"
json_data = JSON.parsefile(json_path)

# Initialize a graph
g = UndirectedGraph()

# Add edges from JSON data
for edge in json_data
    if 2 <= length(edge[2])
        add!(g, edge[1], edge[2][1])
        add!(g, edge[1], edge[2][2])
    end
end

# Plot the graph
embed(g, :spring)
set_vertex_size(g, 1)
draw(g)

savefig("day24.png")

File: misc/2024/day24.R
#!/usr/bin/env Rscript
library(jsonlite)
library(igraph)
library(ggplot2)

j <- fromJSON("2024/day24.json")
r <- Filter(function(l) length(l[[2]]) == 2, j) 
data <- lapply(
    r,
    function(l) {
      list(
        list(from = l[[2]][[1]], to = l[[1]]), list(from = l[[2]][[2]], to = l[[1]])
    )})
data <- unlist(data, recursive = FALSE)
# print(sapply(data, function (l) l[[1]]))
dataframe <- data.frame(
  list(
    from = unlist(sapply(data, function (l) l[[1]])),
    to   = unlist(sapply(data, function (l) l[[2]]))
  )
)
# print(dataframe)
g <- graph_from_data_frame(dataframe, directed = TRUE)
l <- layout_nicely(g)
png("day24-by-R.png")
p <- plot(g, layout = l, edge.arrow.size = .3, vertex.size = 4, vertex.label=NA)
dev.off()
